### PR TITLE
[new release] ocamlgraph_gtk and ocamlgraph (2.0.0)

### PR DESCRIPTION
**Pull request by backtracking**
CHANGES:

- port to dune and opam 2.0
  - :exclamation: opam package now split into two packages: ocamlgraph and ocamlgraph_gtk
  - [WeakTopological] fixed incorrect use of generic hash tables
    (backtracking/ocamlgraph#99, Tomáš Dacík)
  - [Oper] fixed transitive_reduction (backtracking/ocamlgraph#91)
  - fix incorrect uses of polymorphic equality (Steffen Smolka, Boris Yakobowski)
  - [Coloring] fixed generation of OCamlDoc documentation
    (contributed by Earnestly)
  - :exclamation: [Coloring] functions now fail if the graph is directed
  - :exclamation: [Coloring] now uses a single, global exception [NoColoring]
  - [Coloring] new function two_color to 2-color a graph (or fail)
  - :exclamation: [Fixpoint] Take initial labeling of nodes into account (Johannes Kloos)

### CODE REVIEW

**Comment by backtracking** [comment]
Please do not merge, even if CI passes. (We still need to check with the Frama-C team is this is fine for them.)

**Comment by backtracking** [comment]
The Travis CI build failed with OCaml 4.09 with the following error:
```
[ERROR] The compilation of conf-gnomecanvas failed at
        "/usr/local/bin/pkg-config libgnomecanvas-2.0".
```
Is this something I can do anything about?

**Comment by mseri** [comment]
Do you know if conf-gnomecanvas can be made compatible with freebsd?

**Comment by kit-ty-kate** [comment]
> Do you know if conf-gnomecanvas can be made compatible with freebsd?

I just tested locally and there is a `libgnomecanvas-2.0` package. I'll make a separate PR

**Comment by mseri** [comment]
There is only a weird failure on the tests:
```
#=== ERROR while compiling ocamlgraph.2.0.0 ===================================#
# context              2.0.7 | linux/x86_64 | ocaml-base-compiler.4.08.1 | pinned(https://github.com/backtracking/ocamlgraph/releases/download/2.0.0/ocamlgraph-2.0.0.tbz)
# path                 ~/.opam/4.08/.opam-switch/build/ocamlgraph.2.0.0
# command              ~/.opam/4.08/bin/dune build -p ocamlgraph -j 72 @install @runtest
# exit-code            1
# env-file             ~/.opam/log/ocamlgraph-29-ebe0e0.env
# output-file          ~/.opam/log/ocamlgraph-29-ebe0e0.out
### output ###
# File "examples/dune", line 3, characters 23-31:
# 3 |  (libraries graph unix graphics threads))
#                            ^^^^^^^^
# Error: Library "graphics" not found.
# Hint: try:
#   dune external-lib-deps --missing -p ocamlgraph -j 72 @install @runtest
# test_chaotic: all tests succeeded.
# dot: all tests succeeded.
# test_fixpoint: all tests succeeded.
# strat: all tests succeeded.
# test_wto: all tests succeeded.
# test_johnson: all tests succeeded.
# basic: all tests succeeded.
# test_components: all tests succeeded.
# test_bf: all tests succeeded.
# test_topsort alias tests/runtest
# test topsort: all tests succeeded.
#        check alias tests/runtest
# check: all tests succeeded
```

Should we add `graphics` as a `with-test` dependency? Not sure why it fails only on 4.08. If it is not critical, we can just ignore the failure or disable the tests for that case.

**Comment by pascutto** [comment]
It's not clear to me why this fails; it is already added as a test dependency in https://github.com/ocaml/opam-repository/pull/17344/commits/4461cfcda7addf63069644dc9c7bdbd344189031.

**Comment by kit-ty-kate** [comment]
I did a bit of digging here and it seems to be because the `graphics` package does not install any `META` file itself and instead let `ocamlfind` do it (but in the case of OCaml 4.08 ocamlfind wasn't needed) :/ This is a bit weird. @dra27 is that expected?

**Comment by backtracking** [comment]
I added a constraint `<= "1.8.8"` for the 5 packages that did not compile anymore.

File: packages/dose3/dose3.5.0.1/opam [hunk 1]
```diff
@@ -26,7 +26,7 @@ remove: [
 ]
 depends: [
   "ocaml"
-  "ocamlgraph" {>= "1.8.6"}
+  "ocamlgraph" {>= "1.8.6" & <= "1.8.8"}
```
Comment (dra27) [inline packages/dose3/dose3.5.0.1/opam:29]: Wouldn't it be better to allow the possibility of a 1.8.x patch release and make this:
```suggestion
  "ocamlgraph" {>= "1.8.6" & < "2.0.0"}
```
(and others, if you agree)

File: packages/ocamlgraph/ocamlgraph.2.0.0/opam [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  "edge"
+  "persistent"
+  "imperative"
+]
+homepage: "https://github.com/backtracking/ocamlgraph/"
+bug-reports: "https://github.com/backtracking/ocamlgraph/issues/new"
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "stdlib-shims"
+  "dune" {>= "2.0"}
```
Comment (dra27) [inline packages/ocamlgraph/ocamlgraph.2.0.0/opam:22]:
```suggestion
  "dune" {>= "2.0" & !with-test | >= "2.8"}
```
This fixes the issue with the `graphics` package (which at present means `ocamlgraph.2.0.0` can't be installed with `--with-test` on 4.08 and earlier)

File: packages/ocamlgraph_gtk/ocamlgraph_gtk.2.0.0/opam [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+]
+homepage: "https://github.com/backtracking/ocamlgraph/"
+bug-reports: "https://github.com/backtracking/ocamlgraph/issues/new"
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "stdlib-shims"
+  "lablgtk"
+  "conf-gnomecanvas"
+  "ocamlgraph"
+  "dune" {>= "2.0"}
```
Comment (dra27) [inline packages/ocamlgraph_gtk/ocamlgraph_gtk.2.0.0/opam:25]:
```suggestion
  "dune" {>= "2.0" & !with-test | >= "2.8"}
```

**Review by dra27 — COMMENTED** [review]
A couple of constraint suggestions. The alternative, notwithstanding an update to the graphics packages, is to add `"ocamlfind" {with-test}`, but it's a shame to switch to Dune and still actively depend on ocamlfind!

File: packages/dose3/dose3.5.0.1/opam (hunk 1 above)
Comment (backtracking) [reply packages/dose3/dose3.5.0.1/opam]: done

File: packages/ocamlgraph/ocamlgraph.2.0.0/opam (hunk 2 above)
Comment (backtracking) [reply packages/ocamlgraph/ocamlgraph.2.0.0/opam]: done

File: packages/ocamlgraph_gtk/ocamlgraph_gtk.2.0.0/opam (hunk 3 above)
Comment (backtracking) [reply packages/ocamlgraph_gtk/ocamlgraph_gtk.2.0.0/opam]: done.

Many thanks @dra27 for all these suggestions. You have to pardon me for my inexperience with `dune` and `opam` packaging; I'm relatively new to these things.

Comment (dra27) [reply packages/ocamlgraph_gtk/ocamlgraph_gtk.2.0.0/opam]: No problem at all! The graphics/Dune thing is an unusual corner-case anyway. It's lovely to have ocamlgraph ported to build with Dune (having maintained patches for it in opam for quite a while!)

**Comment by mseri** [comment]
I will wait for the big CI to run, but I think this is now good to merge

**Comment by mseri** [comment]
The CI runs look good. Thank you all very much

**Pull request merged**